pkg/portForward/portForward.go

```diff
@@ -75,32 +82,70 @@ func (o *PFClient) StartPortForwarding(
 		return err
 	}
 
-	portsBuf := NewPortWriter(log.GetStdout(), len(portPairsSlice), ceMapping)
+	o.originalErrorHandlers = runtime.ErrorHandlers
+	runtime.ErrorHandlers = append(runtime.ErrorHandlers, func(err error) {
+		if err.Error() == "lost connection to pod" {
+			// Stop the low-level port forwarding
+			// the infinite loop will restart it
+			if o.stopChan == nil {
+				return
+			}
+			o.stopChan <- struct{}{}
+			o.stopChan = make(chan struct{}, 1)
+		}
+	})
+
+	o.isRunning = true
+
+	devstateChan := make(chan error)
 	go func() {
-		err = o.kubernetesClient.SetupPortForwarding(pod, portPairsSlice, portsBuf, errOut, o.stopChan)
-		if err != nil {
-			fmt.Printf("failed to setup port-forwarding: %v\n", err)
+		backo := watch.NewExpBackoff()
+		for {
+			o.finishedChan = make(chan struct{}, 1)
+			portsBuf := NewPortWriter(log.GetStdout(), len(portPairsSlice), ceMapping)
+
+			go func() {
+				portsBuf.Wait()
+				err = o.stateClient.SetForwardedPorts(portsBuf.GetForwardedPorts())
+				if err != nil {
+					err = fmt.Errorf("unable to save forwarded ports to state file: %v", err)
+				}
+				devstateChan <- err
+			}()
+
+			err = o.kubernetesClient.SetupPortForwarding(pod, portPairsSlice, portsBuf, errOut, o.stopChan)
+			if err != nil {
+				fmt.Fprintf(errOut, "Failed to setup port-forwarding: %v\n", err)
+				time.Sleep(backo.Delay())
```

Comment:
In my understanding, in case of an error, we are now retrying with an exponential backoff strategy (capped at 10s by default), which is nice. But would it make sense to give up after a reasonable number of retries rather than keep retrying?
Because I noticed that it is no longer possible to stop the `odo dev` process as long we enter this condition. I simulated this by simulating a Dial error that might be returned by the `kclient.ClientInterface#SetupPortForwarding` implementation in `port_forwarding.go`.

Reply:
I was also wondering. I added some timeout while waiting for the devstate to be written

Reply:
This gives a different behaviour for the first port forwarding and the next restarts: the first one will be timed out, the following ones won't.

But only the first one is blocking (we are waiting for the first port forwarding to be effective before to return the function); the other ones are done in the go routine after the function returned.

Reply:
> This gives a different behaviour for the first port forwarding and the next restarts: the first one will be timed out, the following ones won't.
> 
> But only the first one is blocking (we are waiting for the first port forwarding to be effective before to return the function); the other ones are done in the go routine after the function returned.

This is fine, I think. Once the dev session is running (and ports are forwarded), it is understandable to me that we try to recover from potential errors happening. Thanks.